### DISCUSSION TITLE
Query string functionality with URL

### DISCUSSION THREAD

**Discussion by booni3** [livewire/livewire#5566] [1. Help]
In my app, I have a library route that accesses all Media models in the "media" table. We can search, filter & sort the media library.

Media can also be grouped into collections via a many-to-many relationship in a LibraryCollection model. 

I have a single page within my app that can view both the full library and the library collections. Currently, I am using the query string, but would ideally like to be using route binding as below:

```
// A media library (base URL)
www.example.com/library

// Show recent items
www.example.com/library?filter=recent (current)
www.example.com/library/recent (preferred)

// Show library items in collection 1
www.example.com/library?collection=1 (current)
www.example.com/library-collection/1 (preferred)
```

Using Laravel route binding is cleaner for the user and works more in line with Laravel convention for accessing models i.e. (www.example.com/library-collection/1)

However, I need to be able to switch between the full library and library collections without reloading the page.

I tried writing my own push-history / replace-history methods, but this is a bit messy and `$queryString` handles it so well that it seems the same to do this.

At the moment, I am using the query string, but I would love to be able to use route binding in the same way, with all URL switching and history states managed by livewire.

Is this currently possible using queryString, or will it be possible in V3? I

**Comment by joshhanley** [comment]
@booni3 so that should currently work in V2, but it's not part of the queryString functionality, it should actually just work. So if you browse to say `library-collection/1` and are using route model binding, then updating the model in the component should update the model ID in the URI.

It may also work if the route model binding is an optional model, so you could display the whole library and then select a specific collection, but you'd have to test that. It may need a full page refresh to swap between the library and a specific collection.

V3 will bring better support for things like this though and has SPA mode so it will have that seamless feel to it.

Hope this helps!

**Reply by booni3** [reply]
Thanks for the reply.

Route model binding against a single model works great. The URL changes seamlessly as expected.

My issue is switching between the library and library-collection, which are two separate models that are being accessed via the same url.

From what you have said, that is probably not possible without a full page refresh.

I have it working fine using the query string, but that's not quite what I am aiming for.

**Reply by joshhanley** [reply]
@booni3 no worries!

Yeah possibly doesn't fully work. You may be able to do it now if you use turbolinks/turbo. That will give you have SPA like feel.

**Comment by booni3** [comment]
Thanks. It’s actually less about the spa feel. I have drop-down tree style menus and a few other bits I don’t want changing state when switching between the contexts.


I think v3 might have some features to help with that though.

--- original message ---
On February 21, 2023 at 8:42 PM GMT ***@***.*** wrote:



@booni3 no worries!

Yeah possibly doesn't fully work. You may be able to do it now if you use turbolinks/turbo. That will give you have SPA like feel.

—

Reply to this email directly, view it on GitHub, or unsubscribe.

You are receiving this because you were mentioned.Message ID: ***@***.***>
--- end of original message ---
